kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeOutlinePanel.js

```diff
@@ -0,0 +1,208 @@
+/* eslint-env browser */
+/**
+ * Widget for Narrative Outline Panel
+ * @author David Lyon  <dlyon@lbl.gov>
+ * @public
+ */
+define(['jquery', 'base/js/namespace', 'kbwidget', 'kbaseNarrativeControlPanel'], (
+    $,
+    Jupyter,
+    KBWidget,
+    ControlPanel
+) => {
+    'use strict';
+    return new KBWidget({
+        name: 'kbaseNarrativeOutlinePanel',
+        parent: ControlPanel,
+        version: '1.0.0',
+        options: {
+            /** If true, all markdown H1-H3 will be outline items
+             * If false, only the first header will be an outline item */
+            showAllHeaders: false,
+            /** Header height for visibility & scroll functionality */
+            headerHeight: 80,
+            /** Refresh throttle rate in ms */
+            refreshRate: 100,
+        },
+
+        isActive: false,
+        body: undefined,
+        drawnState: '', // string representing the items drawn used to eliminate excess dom manipulation
+
+        init: function (options) {
+            this._super(options);
+            this.body = $(this.body());
+
+            const refresh = this.refresh.bind(this);
+            // Trigger refresh on the following events
+            $([Jupyter.events]).on('notebook_loaded.Notebook', refresh);
+            $([Jupyter.events]).on('notebook_saved.Notebook', refresh);
+            $([Jupyter.events]).on('notebook_renamed.Notebook', refresh);
+            $([Jupyter.events]).on('command_mode.Notebook', refresh); // Clicking out of a cell
+            $([Jupyter.events]).on('create.Cell', refresh);
+            $([Jupyter.events]).on('delete.Cell', refresh);
+            $([Jupyter.events]).on('select.Cell', refresh); // Triggers on cell move
+            $('#notebook-container').scroll(refresh);
+
+            // Hide panel header
+            this.$elem.find('.kb-title').hide();
+
+            return this;
+        },
+
+        activate: function () {
+            // Render when the panel is activated
+            this.renderOutline();
+            this.isActive = true;
+        },
+
+        deactivate: function () {
+            this.isActive = false;
```

Comment:
should the widget contents be removed when it is deactivated?

Reply:
I'm not sure the exact lifecycle of kbwidgets if im honest, but I think leaving the dom contents is best, then it doesnt have to rerender if nothing changes between deactivation and reactivation (if that happens?).

Reply:
When the tab gets activated by clicking its header, the `activate()` function gets called. So it looks like there's a rerender here. It seems to not be a heavy step, so I think it'll be fine for now.